### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by acowley** [gibiansky/IHaskell#647]
I have a rather large Org document in emacs that I tie into IHaskell for source blocks. It's a really cool setup, but I do have one issue. 

One can export the document through various backends, in this case I'm exporting my document to a reveal.js slide show. Since the IHaskell bits are all in one session, the export procedure evaluates every block from the top to make sure the output is consistent, but I find that after two or three exports, my IHaskell process has crashed. I then need to make sure the Emacs bits are killed because they don't do anything intelligent in response to the IHaskell process dying. At that point I can run another export operation, and IHaskell is restarted and all is well until the next time.

This is the only error message I can surface:
`ihaskell: getWorkingDirectory: resource exhausted (Too many open files)`

The notebook does a fair bit of importing, and outputs a dozen or more images, so there is quite a bit of file IO.

TL;DR: Does anyone have a guess as to how we might reduce the number of file handles IHaskell holds onto?

**Comment by gibiansky** [comment]
I've seen this error a few times myself. I don't know where it comes from. 

A first step to debug this is to probably look at the `/proc` filesystem or `lsof` to figure out *what* file handles are open. This could be due to a few things and would be necessary to figure out which of them causes it. If the file handles are open to image files, it would be the IHaskell extensions; if it's to Haskell files or the like, it could be due to the `hlint` support. Nothing else comes to mind immediately, but it's important to identify *which* part of IHaskell leaks file handles.

**Comment by acowley** [comment]
I took a quick look at the OS X Activity Monitor process info tab for IHaskell, and all those `/dev/ttys005` file descriptors do look suspicious. Here is the [full list of open files and ports](http://lpaste.net/5976118414500954112) after two exports of my document.

**Comment by gibiansky** [comment]
Huh. I have no idea what those might be; they're Unix serial devices / TTYs?

I wonder if, say, every time a cell is evaluated, a new `/dev/ttys005` is opened... Or perhaps every time a cell that generates an image is evaluted. For example, perhaps these are issues being triggered in `hlint` (which gets run a ~once per cell), or GHC, or IHaskell itself.

**Comment by acowley** [comment]
Is it correct that there is no line,

`voidpf "IHaskellIO.closeFd %s" readVariable` in `postStmts`
in `capturedEval` in `Evaluate.hs`? Right [here](https://github.com/gibiansky/IHaskell/blob/master/src/IHaskell/Eval/Evaluate.hs#L1101).

**Comment by acowley** [comment]
Is there an easy way to temporarily disable `hlint` so I can see if the overall behavior changes?

**Comment by gibiansky** [comment]
`:hlint off` I think will do so.

**Comment by gibiansky** [comment]
Also `:help` will give you the right command if that's not it

**Comment by gibiansky** [comment]
And yeah, good catch, closing `readVariable` is necessary and I think not done. That may be the issue!

**Comment by acowley** [comment]
Closing `readVariable` makes it so Jupyter doesn't get output from IHaskell :/

I don't know why that is, but just running `jupyter console --kernel haskell` and entering an expression produces no output. So, that FD might be problematic, but closing it there does no good.

**Comment by gibiansky** [comment]
You may need to close it later, after it has been read from? We spawn a separate thread to read from that file handle, which loops and reads from that handle until the computation is done. See the code `-- Once context is finalized, reading can finish. Wait for reading to finish to that the output`; after we `takeMVar finishedReading` it should be safe to close that. So instead of adding it to `postStmts` (which get evaluated *before* `takeMVar finishedReading`, run it after `takeMVar finishedReading`.

Basically it's possible that the issue is as follows:

- We start the loop in another thread that gets delayed. However it does not start running yet; the scheduler has not scheduled it yet. Especially if we only have one OS thread.
-  We run the preparation statements and the main cell content and the post statement, and write to the `completed` `MVar`. 
- We check for `finishedReading`, but it's not yet written to, so we block until reading ends.
- The scheduler schedules our `loop` thread, which does the reading. It starts reading. HOWEVER since `postStmts` have been run, the handle is already closed! It fails at reading.
- We now continue onwards, having read nothing.

If we only close the `readVariable` *after* we are sure we are done reading, maybe the issue will be fixed?

**Comment by acowley** [comment]
I tried this, and it broke output to file for images. I don't have any IHaskell error messages, but on the emacs side there was no output to be piped to a file.

**Comment by gibiansky** [comment]
That's very surprising. `readVariable` has, I think,  more or less nothing to do with image writing. Since it causes problems on images, it may be a timing issue with the threading, since images presumably take a long time to write :-/ gnarly issue

**Comment by acowley** [comment]
Yeah, it's pretty black-and-white for me: if I evaluate the closeFd on `readVariable` after `takeMVar finishedReading` my very first image output fails. If I leave the IHaskell side alone, it works fine.

Btw, `:option no-lint` didn't solve this problem, but it *did* have the very nice effect of massively reducing memory retention. Peak memory usage is unchanged, but with linting turned off, the memory is freed when IHaskell goes idle, whereas before everything was held.

**Comment by nickie** [comment]
Has this issue been fixed?  I'm getting the same behaviour with Jupyter console 6.0.0, IHaskell 0.10.0.2.

**Comment by vaibhavsagar** [comment]
Do you have a small example to help me reproduce this behaviour?

**Comment by nickie** [comment]
I'm attaching a Haskell Jupyter notebook (zipped) importing `System.Random` and calling `randomIO` 50 times.  After the 47th call, it reports:

```
ihaskell: /dev/urandom: openFd: resource exhausted (Too many open files)
```

If that's not the case on your machine, you may have to introduce a few more `randomIO` calls.
[open-files-bug.zip](https://github.com/gibiansky/IHaskell/files/3848844/open-files-bug.zip)

**Comment by nickie** [comment]
Should I make a new issue for this?

**Comment by vaibhavsagar** [comment]
I'm not sure this is an issue with IHaskell. Does this work in GHCi?

**Comment by vaibhavsagar** [comment]
I tried this with 89 cells and could not reproduce the issue you're experiencing.

**Comment by nickie** [comment]
I cannot reproduce it either with the latest version of `master` on Linux (aee398a82fb87146c39902dbefea723d1799292e).  It's either a bug that manifests only on Mac or something that got fixed recently (after e15dd63d2efdcfda599dc022732b33589005a6d5).  I'll check again.

**Comment by nickie** [comment]
It must be a Mac issue because e15dd63 on Linux works OK.  In contrary, the latest aee398a does not work on Mac.
